core/src/main/java/org/apache/calcite/rex/RexSimplify.java

```diff
@@ -2331,17 +2331,20 @@ private void verify(RexNode before, RexNode simplified, RexUnknownAs unknownAs)
 
   private RexNode simplifySearch(RexCall call, RexUnknownAs unknownAs) {
     assert call.getKind() == SqlKind.SEARCH;
-    final RexNode a = simplify(call.getOperands().get(0), unknownAs);
+    final RexNode operand = call.getOperands().get(0);
+    final RexNode simplifiedOperand = simplify(operand, unknownAs);
+    final boolean operandUnchanged = operand.equals(simplifiedOperand);
```

Comment:
should this be ==?

Reply:
Hmm, I am not really sure if the same object will be returned when the operand cannot be simplified any further. But I see that `RexCall.equals` checks this in the beginning:
```
    if (this == o) {
      return true;
    }
```
So I think call to `.equals` is safer and without being too inefficient.

But do let me know what your thoughts are regarding this.